src/sources/models.js

```diff
@@ -3,14 +3,67 @@ const {fetch} = require("../fetch");
 const {NoResourcePathError} = require("../exceptions");
 const utils = require("../utils");
 
-
-/* These Source, Dataset, and Narrative classes contain information to map an
- * array of dataset/narrative path parts onto a URL.  Source selection and
- * dataset path aliasing (/flu → /flu/seasonal/h3n2/ha/3y) is handled in
- * utils/prefix.parsePrefix().
+/* The model classes here are the base classes for the classes defined in
+ * ./core.js, ./community.js, ./groups.js, etc.
+ *
+ * Sources represent a remote HTTP data source.  Sources contain Datasets and
+ * Narratives (both Resources).  Datasets and Narratives have Subresources
```

Comment:
And from a slightly wider point of view, the current situation instantiates objects each time the endpoint is hit. This is nice as we can check autn each time etc. But there are other use-cases where a cache / stored state is necessary - e.g. we don't want to recompute the available datasets for a particular source each time a request is made. We currently get around this in different ways: using `globals` and a  `MetaGroupSource` which essentially stores the `getAvailable` data for each group. 

Is there a high level way where newly instantiated classes can have access to a cache (or similar) of data? Or could we cache at the level of the request itself (e.g. `fetch(source.available.subresource("???").url())` doesn't actually make a network response but we maintain a periodically updated cache).

Reply:
Currently the Charon endpoints `getSourceInfo` and `getAvailable` use methods on Source classes which directly return data instead of returning a reference to data (e.g. a URL), which is a design that still makes sense to me. I'm not sure this needs to change anytime soon, but if we had reasons to switch that information to static files for decoupling/caching reasons, then I can imagine using new Resource subclasses for them.

Newly instantiated instances of these classes can still access caches, and there are a number of different options for caching that can be mixed and matched as needed including per-process, in-memory caching like we do now, or external, shared caching in something like Redis, or expanding our network caching (with make-fetch-happen).